crates/telio-wg/src/pq.rs

```diff
@@ -0,0 +1,331 @@
+const REMOTE_IP: Ipv4Addr = Ipv4Addr::new(10, 5, 0, 1);
+const HANDSHAKE_VERSION: u32 = 1;
+const CIPHERTEXT_LEN: u32 = kyber768::ciphertext_bytes() as _;
+
+const IPV4_HEADER_LEN: usize = 20;
+const UDP_HEADER_LEN: usize = 8;
+
+/// The PQ module error type
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// IO error
+    #[error("IO: {0:?}")]
+    Io(#[from] io::Error),
+    /// Generic unrecoverable error
+    #[error("Generic: {0}")]
+    Generic(String),
+}
+
+impl From<String> for Error {
+    fn from(value: String) -> Self {
+        Self::Generic(value)
+    }
+}
+
+impl From<&str> for Error {
+    fn from(value: &str) -> Self {
+        Self::Generic(value.to_string())
+    }
+}
+
+impl From<tokio::time::error::Elapsed> for Error {
+    fn from(value: tokio::time::error::Elapsed) -> Self {
+        Self::Io(io::Error::new(io::ErrorKind::TimedOut, value))
+    }
+}
+
+/// Post quantum keys retrived from hanshake
+pub struct PqKeys {
+    /// Kyber shared secret
+    pq_shared: kyber768::SharedSecret,
+    /// X25519 secret key
+    wg_secret: x25519_dalek::StaticSecret,
+}
+
+impl PqKeys {
+    /// Retrieve the PQ preshared key
+    pub fn pq_shared(&self) -> telio_crypto::PresharedKey {
+        #[allow(clippy::expect_used)]
+        telio_crypto::PresharedKey(
+            self.pq_shared
+                .as_bytes()
+                .try_into()
+                .expect("Preshared WG keys should have the same lenght as PQ preshared key"),
+        )
+    }
+
+    /// Retrieve the WG private key
+    pub fn wg_secret(&self) -> telio_crypto::SecretKey {
+        telio_crypto::SecretKey::new(self.wg_secret.to_bytes())
+    }
+}
+
+/// Get PQ keys from the VPN server
+/// The packet sequence is as follows
+///
+/// 1) C -> S             : Client sends WG hansake packet                           
+/// 2) S -> C             : Server responds to handsake                              
+/// 3) C -> S (in tunnel) : Client sends a UDP PQ GET packet with keys               
+/// 4) S -> C (in tunnel) : Server responds with UDP packet containing the ciphertext
+///
+/// The shared key is established which should be used as,
+/// a WG preshared key for subsequent connection
+pub async fn fetch_keys(
+    endpoint: impl ToSocketAddrs,
+    secret: &telio_crypto::SecretKey,
+    peers_pubkey: &telio_crypto::PublicKey,
+) -> Result<PqKeys, Error> {
+    let (tunn, sock) = handshake(endpoint, secret, peers_pubkey).await?;
+
+    let mut rng = rand::rngs::StdRng::from_entropy();
+
+    // Generate keys
+    let wg_secret = x25519_dalek::StaticSecret::random_from_rng(&mut rng);
+    let (pq_public, pq_secret) = kyber768::keypair();
+    let wg_public = x25519_dalek::PublicKey::from(&wg_secret);
+
+    let mut pkgbuf = Vec::with_capacity(1024 * 4); // 4 KiB
+
+    pkgbuf.resize(IPV4_HEADER_LEN + UDP_HEADER_LEN, 0);
+    push_get_method_udp_payload(&mut pkgbuf, &wg_public, &pq_public);
+    fill_get_packet_headers(pkgbuf.as_mut_slice(), &mut rng);
```

Comment:
Much nicer! But isn't it a bit C-style? Maybe we could just return this `pkgbuf` from a function and avoid making it mutable?

Reply:
Let, me try

Reply:
done